Models and Routes

ttpham0111 opened the pull request:
Has been tested for successful requests for all requests on google docs. Some errors will still send a success message but will not perform any actions on the database.

Commit:
Created outline for user.js and linked map.js and user.js to server.js

Commit:
Created outline for map.js

Commit:
Updated outline of map.js and user.js

Commit:
Updated user.js

banhathome committed:
Added database config file.

banhathome committed:
Added createTag method
Added temporary code to show how to call a model in server.js.

Commit:
Updated map.js and user.js to match expected functionality for v1

Commit:
Merge branch 'models-1' of https://github.com/DevelopersGuild/daze-backend into Tleaf

Commit:
Initial commit for database

Commit:
Created structure of tag and user models

Commit:
Updated map and user routes to match model methods

Commit:
Updated to handle token creation/refresh/deletion

Commit:
Stylistic change from tabs to spaces

Commit:
Reformatted to look cleaner

Commit:
Changed names of body.token to tokenID to avoid conflicts with callback

Commit:
Removed sample usage of models

Commit:
Fixed error with path for require(path/to/db)

Commit:
Implemented functionality with session models

Commit:
Fixed error with require paths

Commit:
Modified to work with new user model methods.

Commit:
Removed test variables

Commit:
Initial commit with Schema and methods included

Commit:
Handled errors and failed queries resulting in null documents

Commit:
Handled errors and failed queries resulting in null documents

Commit:
Added location to schema

Commit:
Handled errors and null document queries

banhathome commented on models/sessions.js:
```diff
@@ -0,0 +1,124 @@
+			callback(err, null);
+			return;
+		}
+		if (session)
```
Please use brackets

banhathome commented on models/sessions.js:
```diff
@@ -0,0 +1,124 @@
+		}
+		if (session)
+			callback(null, session.token);
+		else
```
Please use brackets

banhathome commented on models/sessions.js:
```diff
@@ -0,0 +1,124 @@
+			callback({
+				code	: 400,
+				message : 'The session does not exist.'
+			}, null);
```
Can we change this to the following instead:

```
if (err) {
...
} else if (!session) {
...
} else {
...
}

```

It's a bit awkward to have the success instance sandwiched between two error instances.

banhathome replied:
You don't need to specify that the second argument is null. You can just leave empty

ttpham0111 replied:
I left null to remind me that the callback takes 2 arguments and the second one wasn't found in that if block.

banhathome commented on models/sessions.js:
```diff
@@ -0,0 +1,124 @@
+			callback({
+				code	: 400,
+				message : 'The session does not exists.'
+			}, null);
```
See above

banhathome commented on models/sessions.js:
```diff
@@ -0,0 +1,124 @@
+			callback({
+				code	: 400,
+				message : 'Session creation failed.'
+			}, null);
```
See above

banhathome commented on models/sessions.js:
```diff
@@ -0,0 +1,124 @@
+			return;
+		}
+		if (session)
+			callback(null, session);
```
Maybe we should only return the session token here, so `callback(null, session.token)`

ttpham0111 replied:
In the google docs we need to return both the token string and the TTL, which both belong to the session. We can do a callback with 3 arguments? I still need to figure out the TTL for the response.

banhathome commented on models/sessions.js:
```diff
@@ -0,0 +1,124 @@
+	SessionMongoModel.findOne({ userId : userId }, function(err, session) {
+		if (err) {
+			// TODO: Error message?
+			callback(err, null);
```
You don't need to specify that the second argument is null. You can leave it empty, so just `callback(err)`

banhathome commented on models/sessions.js:
```diff
@@ -0,0 +1,124 @@
+
+// Takes a token string and deletes the session with that string from MongoDB.
+function deleteSession(clientToken, callback) {
+	SessionMongoModel.findOneAndRemove({ token : clientToken }, function(err, session) {
```
This line is too long. Split it up into two lines:

```
SessionMongoModel.findOneAndRemove({ token : clientToken }, 
  function(err, session) {
```

banhathome commented on models/sessions.js:
```diff
@@ -0,0 +1,124 @@
+}
+
+// Takes a token string and deletes the session with that string from MongoDB.
+function deleteSession(clientToken, callback) {
```
Nitpick: `destroySession` would be a better name because "destroying sessions" is usually more commonly used than "deleting sessions"

banhathome commented on models/sessions.js:
```diff
@@ -0,0 +1,124 @@
+			callback(err);
+			return;
+		}
+		if (session)
```
Even though the method you invoked removes the session from the db, it still returns the `session` as a valid second argument, meaning that `session` is valid even if it was successful.

ttpham0111 replied:
Will that session actually be pointing to something, or is it null?

ttpham0111 replied:
It isn't null even when it destroyed it. Weird, oh well.

banhathome commented on models/sessions.js:
```diff
@@ -0,0 +1,124 @@
+				message : 'Session deletion failed.'
+			});
+		else 
+			callback(null);
```
See above.

banhathome commented on models/sessions.js:
```diff
@@ -0,0 +1,124 @@
+// Takes a token string and update the createdAt field to Date.now()
+function refreshSession(clientToken, callback) {
+	SessionMongoModel.findOneAndUpdate({ token : clientToken },
+									   {$set : { createdAt : Date.now }},
```
The indentation of this line is off.

banhathome replied:
This should also be `Date.now()`

banhathome commented on models/sessions.js:
```diff
@@ -0,0 +1,124 @@
+}
+
+// Takes a token string and update the createdAt field to Date.now()
+function refreshSession(clientToken, callback) {
```
Can we instead destroy the old session, if exists, and create a new one?

banhathome commented on models/sessions.js:
```diff
@@ -0,0 +1,124 @@
+function refreshSession(clientToken, callback) {
+	SessionMongoModel.findOneAndUpdate({ token : clientToken },
+									   {$set : { createdAt : Date.now }},
+									   function(err, session) {
```
The indentation of this line is off

banhathome replied:
You should handle the error that occurs in the event that Mongo doesn't return an error, but the session wasn't found.

banhathome commented on models/sessions.js:
```diff
@@ -0,0 +1,124 @@
+	refresh     : refreshSession
+};
+
+module.exports = SessionModel;
```
Github is complaining that you didn't leave an empty line at the end of your file. This isn't too important, but not ending your file with an empty line can cause errors, specifically when we're minifying code.

banhathome commented on models/tags.js:
```diff
@@ -0,0 +1,75 @@
+    // Expiration set to Date.now() + TTL
+    expireAt       : {type: Date  , required: true, expireAfterSeconds: 0 },
+
+    // TODO: What is score???
```
I'm not sure either

Commit:
Added brackets for if/else statements to improve readability. Modified destroySession to create a new sessions instead of updating

Commit:
Changed deleteSession to destroySession to coincide with popular convention.

Commit:
Added empty line at the end to avoid problems with GitHub and minifying code.

Commit:
Fixed async problems with createUser method

Commit:
Added brackets to if/else scope. Modified to use if/else instead of returns

Commit:
Removed null document check on document deletion.

Commit:
Modified to return session from model to routes instead of token

ttpham0111 commented:
Tested all codes so far, still working properly :+1:

Commit:
TTL added to response object - unfinished

Commit:
Added empty line at end of file to avoid issues with GitHub

banhathome commented on models/users.js:
```diff
@@ -0,0 +1,211 @@
+    });
+
+    // Check if newUser.username is already in UserMongoModel
+    UserMongoModel.find({ username : username }).limit(1).exec(function(err, user) {
```
use `UserMongoModel.findOne`

banhathome replied:
Also, mongo is case-sensitive, so if the user creates their account as `Testing` and tries to log in as `testinG`, it'll fail.

We should instead change the find to `{ usernameLower: username.toLowerCase()}`

ttpham0111 replied:
collection.find is apparently a whole lot faster because it returns a cursor instead of a document, so it's more ideal for checking if a document exists... Refer to: https://blog.serverdensity.com/checking-if-a-document-exists-mongodb-slow-findone-vs-find/

banhathome replied:
Interesting. Let's keep it how it is then.

banhathome replied:
Hey @tuan91187dsf, I just tested it out and `.find()` returns the whole document, as opposed to just a cursor. This might be because we're not using Mongo directly; instead we're using Mongoose, which doesn't support cursors. This means that `.find().limit(1)` behaves exactly like `.findOne()`, so we should go with the latter for simplicity's sake.

ttpham0111 replied:
@chrisngobanh You're right, moongoose converts the cursor to an array, so it's actually slower to do a find()...

banhathome commented on models/users.js:
```diff
@@ -0,0 +1,211 @@
+
+    // Check if newUser.username is already in UserMongoModel
+    UserMongoModel.find({ username : username }).limit(1).exec(function(err, user) {
+        if (user.length) {
```
Should this be `if (user) {` instead?

ttpham0111 replied:
Because find returns a cursor, we have to check if the cursor holds anything. See above.

banhathome replied:
See above.

banhathome commented on models/users.js:
```diff
@@ -0,0 +1,211 @@
+        } else {
+
+            // Check if newUser.email is already in UserMongoModel
+            UserMongoModel.find({ email : email }).limit(1).exec(function(err, user) {
```
Use `UserMongoModel.findOne`

banhathome replied:
We should change the find to `{email: email.toLowerCase()}`

banhathome commented on models/users.js:
```diff
@@ -0,0 +1,211 @@
+// Takes username, password, email and saves a new user to MongoDB
+function createUser(username, password, email, callback) {
+    var newUser = new UserMongoModel({
+        username : username,
```
Mongo is case-sensitive, so you should include a `usernameLower` field as well so we can log the user in regardless of how they type their username, while maintaining the way they originally wanted to make their username.

banhathome commented on routes/user.js:
```diff
@@ -0,0 +1,238 @@
+                code    : 200,
+                message : 'Account successfully created.',
+                token   : session.token,
+                ttl     : session.createdAt.setDate(session.createdAt.getDate() + 7).toString()
```
This should instead be `ttl: session.createdAt + 604800000 // 604800000 is 7D in milliseconds`

Commit:
Added .toLowerCase() to queries to allow accidental uppercase in login fields

Commit:
Modified to use findOne instead of find because mongoose converts cursors to arrays automatically

Commit:
Added usernameLower to model and all username checks rely on usernameLower

Commit:
Converted to 2 space tabs to improve readibility. routes/user.js response object now returns session ttl as UNIX time.

banhathome commented on models/sessions.js:
```diff
@@ -0,0 +1,133 @@
+  SessionMongoModel.findOne({ userId : userId.toLowerCase() }, function(err, session) {
+    if (err) {
+      // TODO: Error message?
+      callback(err, null);
```
You don't need to make the second argument `null`

banhathome commented on models/sessions.js:
```diff
@@ -0,0 +1,133 @@
+  newSession.save(function (err, session) {
+    if (err) {
+      // TODO: Error message?
+      callback(err, null);
```
You don't need to make the second argument `null`

chrisngobanh commented on models/sessions.js:
```diff
@@ -0,0 +1,133 @@
+      callback(err, null);
+    } else if (!session) {
+      callback({
+        code    : 400,
```
This would be a 500 instead.

chrisngobanh replied:
Actually, you don't need to check to see if `session` exists because it should always exist if `err` doesn't exist.

banhathome commented on models/sessions.js:
```diff
@@ -0,0 +1,133 @@
+
+// Takes a userId as an argument and returns the token string to the callback
+function findSessionByUserId(userId, callback) {
+  SessionMongoModel.findOne({ userId : userId.toLowerCase() }, function(err, session) {
```
Line is too long

chrisngobanh commented on models/sessions.js:
```diff
@@ -0,0 +1,133 @@
+    if (err) {
+      // TODO: Error message?
+      callback(err);
+    } else if (!session) {
```
See above.

banhathome commented on models/sessions.js:
```diff
@@ -0,0 +1,133 @@
+// Takes a token string, deletes the associated session, and makes a new one,
+// returning the new token to the callback
+function refreshSession(clientToken, callback) {
+  SessionMongoModel.findOne({ token : clientToken },
```
You probably don't need to check if the token exists because destroySession should fail if it doesn't exist.

banhathome commented on models/sessions.js:
```diff
@@ -0,0 +1,133 @@
+        if (err) {
+          callback(err);
+        } else {
+          destroySession(clientToken, function(err) {
```
Use async to avoid callback hell

banhathome commented on models/sessions.js:
```diff
@@ -0,0 +1,133 @@
+
+// Takes a token string, deletes the associated session, and makes a new one,
+// returning the new token to the callback
+function refreshSession(clientToken, callback) {
```
Maybe we should handle "refreshing sessions" in user.reauthenticate.

banhathome commented on models/tags.js:
```diff
@@ -0,0 +1,75 @@
+});
+
+// Creates a collection named tags in MongoDB
+var TagMongoModel = Db.model('tags', TagSchema);
```
We should rename this to `markers` and `markerSchema`, etc.

banhathome commented on models/tags.js:
```diff
@@ -0,0 +1,75 @@
+  newTag.save(function(err, tag) {
+    if (err) {
+      callback(err);
+    } else if (!tag) {
```
`tag` will never be undefined if `err` is undefined.

banhathome commented on models/users.js:
```diff
@@ -0,0 +1,216 @@
+  });
+
+  // Check if newUser.username is already in UserMongoModel
+  UserMongoModel.findOne({ usernameLower : username.toLowerCase() }, function(err, user) {
```
Line is too long

banhathome commented on models/users.js:
```diff
@@ -0,0 +1,216 @@
+  var newUser = new UserMongoModel({
+    username      : username,
+    usernameLower : username.toLowerCase(),
+    email         : email,
```
This should be `email: email.toLowerCase()`

banhathome commented on models/users.js:
```diff
@@ -0,0 +1,216 @@
+
+// Takes username, password, email and saves a new user to MongoDB
+function createUser(username, password, email, callback) {
+  var newUser = new UserMongoModel({
```
We should define newUser after checking if the username and email are not taken.

banhathome commented on models/users.js:
```diff
@@ -0,0 +1,216 @@
+            if (err) {
+              // TODO: Error message?
+              callback(err);
+            } else if (!user) {
```
`user` will never be undefined if `err` is undefined.

banhathome commented on models/users.js:
```diff
@@ -0,0 +1,216 @@
+
+// Takes a token string and deletes the associated user and session from MongoDB
+function deleteUser(clientToken, callback) {
+  Session.findUser(clientToken, function(err, userId) {
```
You probably don't need to do this.

banhathome commented on models/users.js:
```diff
@@ -0,0 +1,216 @@
+        } else {
+
+          // Removes session from sessions collection
+          Session.destroy(clientToken, function(err) {
```
We should have a `Session.destroyAll` method instead to destroy all the user's sessions if they delete their account.

banhathome commented on models/users.js:
```diff
@@ -0,0 +1,216 @@
+    } else {
+
+      // Remove user from UserMongoModel
+      UserMongoModel.findByIdAndRemove(userId.toLowerCase(), function(err, user) {
```
userId is an objectId, so you shouldn't do a `.toLowerCase()` on it

banhathome replied:
Line too long

banhathome commented on models/users.js:
```diff
@@ -0,0 +1,216 @@
+  });
+}
+
+// Takes a token string, password, new password and updates the user collection with new password
```
Line too long

banhathome commented on models/users.js:
```diff
@@ -0,0 +1,216 @@
+      // Error message handled by session model
+      callback(err);
+    } else {
+      UserMongoModel.findById(userId.toLowerCase(), function(err, user) {
```
See above.

banhathome commented on models/users.js:
```diff
@@ -0,0 +1,216 @@
+        } else {
+
+          // Verify password
+          if (user.password != oldPassword) {
```
`use !==` instead of `!=`

Commit:
Removed Session.refresh method, to be handled in user model

Commit:
Converted to async waterfalls and parallels

Commit:
Replaced equality check for password change with a deep equality check. Email addresses are now converted to lowercase before being stored

Commit:
Added missing ; to statements

Commit:
Removed redundant null object check for saving new user to database

Commit:
All instances of "tag" replaced with "marker"

Commit:
Removed embeded parallel async to fix a bug with account creation

Commit:
Modified to handle Session.expireAt
. Date additions still does not work properly.

banhathome commented on config/database.js:
```diff
@@ -0,0 +1,29 @@
+
+// Tell Mongoose where our db is located. In this case, we are getting
+// the environment variable SERVER_IP which should be saved in the bash
+// in the folder DG-Forum
```
Update this comment

chrisngobanh commented on routes/user.js:
```diff
@@ -0,0 +1,237 @@
+      }
+      res.send({
+        code    : 200,
+        message : 'Login successful.',
```
Change this message to 'Reauthentication Successful', or whatever you think would be appropriate.

banhathome commented on routes/user.js:
```diff
@@ -0,0 +1,237 @@
+  }
+
+  function authenticate(req, res) {
+    var usernameEmail = validator.toString(validator.escape(req.body.usernameEmail));
```
Nitpick: Line too long

chrisngobanh commented on routes/user.js:
```diff
@@ -0,0 +1,237 @@
+        code    : 200,
+        message : 'Login successful.',
+        token   : session.token,
+        ttl     : session.expireAt
```
should this instead be `ttl     : session.expireAt - Date.now()`

chrisngobanh commented on routes/user.js:
```diff
@@ -0,0 +1,237 @@
+        code    : 200,
+        message : 'Login successful.',
+        token   : session.token,
+        ttl     : session.expireAt - Date.now()
```
This returns a negative Value.

banhathome commented on models/sessions.js:
```diff
@@ -0,0 +1,101 @@
+  createdAt : { type : Date, default : Date.now, expires : '7d' },
+
+  // Defaults to the current time + 7 days
+  expireAt  : { type : Date, default : Date(Date.now + 604800000) },
```
This should be `expireAt  : { type : Date, default : Date.now + 604800000 },` instead

banhathome replied:
Don't do math here

banhathome commented on models/users.js:
```diff
@@ -0,0 +1,271 @@
+            message : 'Incorrect Password'
+          });
+        }
+        next(null, userId);
```
This should be `else { next(null, userId); }`

banhathome commented on models/users.js:
```diff
@@ -0,0 +1,271 @@
+            code    : 400,
+            message : 'User does not exist.'
+          });
+        } else if (user.password != oldPassword) {
```
use `!==` instead of `!=`

banhathome commented on models/tags.js:
```diff
@@ -0,0 +1,75 @@
+  tag         : {type: String, required: true},
+
+  // Username of owner of the tag
+  owner       : {type: String, required: true},
```
Change to MongoId

banhathome committed:
Finished backend according to specifications.
Fixed a lot of bugs.
Refactored a lot.
Introduced the usage of bcrypt to encrypt passwords.
Removed input sanitation for the time being.
